common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java

```diff
@@ -264,6 +265,8 @@ private void checkAuth(TransportClient client, String appId) {
     private final Timer registerExecutorRequestLatencyMillis = new Timer();
     // Time latency for processing finalize shuffle merge request latency in ms
     private final Timer finalizeShuffleMergeLatencyMillis = new Timer();
+    // Block transfer rate in blocks per second
```

Comment:
Is this valid when we do `getContinuousBlocksData`?

Reply:
This is a great point, thanks Dongjoon!

Currently a single batch fetch will be considered as one block fetch by this metric, regardless of how many blocks are fetched in the block. For our purposes, this is what we're interested in, since a big part of what we want to understand with this is the number of random reads we're submitting.

However I see value in breaking it out as the actual number of blocks as well. Perhaps we can have two metrics, `blockTransferRate` and `blockFetchRequestRate`. In the case of non-batch fetches they will be the same, and with batch fetches the `blockTransferRate` will be higher than the `blockFetchRequestRate`. WDYT?

Reply:
On second thought, I don't like `blockFetchRequestRate` since I think `FetchRequest` sounds like it would be at the level of a `OpenBlocks`/`FetchShuffleBlocks` message vs. the individual response level.

Maybe `blockTransferMessageRate`? Or `blockBufferTransferRate`?

Reply:
Pushed up a new commit with `blockTransferMessageRate` to demonstrate the idea, let me know what you think.

Reply:
@dongjoon-hyun how does the new code look to you?

Reply:
@mridulm I would also be curious about your thoughts on the new metrics.

Reply:
In spark read side, we treat a single read of `ShuffleBlockBatchId` as a single block read (from metric point of view).
A [recent discussion](https://github.com/apache/spark/pull/32140#discussion_r652733447) of this in push based shuffle here for context (see `ShuffleBlockBatchId` in `ShuffleBlockFetcherIterator`).

Given that, I am fine with treating a batch read as a single read - given it would be contiguous (effectively, similar to reading a 'large' block).

Thoughts @dongjoon-hyun  ?

Reply:
Thanks for that context @mridulm! To clarify, are you suggesting we ditch the new proposal and only maintain the old version, which would treat a batch fetch as a single block transfer?

From the comments on #32140, it seems like @otterc was indicating that there's a future plan to add metrics on the push-based shuffle side that also break out the merged chunks in terms of number of underlying blocks. Should we do the same here (which is basically my new proposal)?

Reply:
That would be a new metric, if introduced, in read side.
Currently, what is tracked treats it as a single request - that request could be a block fetch or a single fetch for a large block or merged block read.

Reply:
Okay. I am pretty neutral here, happy to go either way. @dongjoon-hyun or @otterc , any thoughts?

Reply:
I don't have anything much to add here . Would be okay either ways.

Reply:
Would definitely want to hear @dongjoon-hyun's thoughts here, given he has context !
Also, +CC @Ngone51 who is helping review the client side for push based shuffle and has looked at similar codepaths recently.

Reply:
I am assuming there are no further suggestions to resolve here ... we can do a follow up in case there is a change in behavior required for 3.2
+CC @dongjoon-hyun